chisel/src/dispatcher.rs

```diff
@@ -0,0 +1,574 @@
+
+/// Chisel input dispatcher
+#[derive(Debug)]
+pub struct ChiselDisptacher {
+    /// The status of the previous dispatch
+    pub errored: bool,
+    /// A Chisel Session
+    pub session: ChiselSession,
+}
+
+/// Chisel dispatch result variants
+#[derive(Debug)]
+pub enum DispatchResult {
+    /// A Generic Dispatch Success
+    Success(Option<String>),
+    /// A Generic Failure
+    Failure(Option<String>),
+    /// A successful ChiselCommand Execution
+    CommandSuccess(Option<String>),
+    /// A failure to parse a Chisel Command
+    UnrecognizedCommand(Box<dyn Error>),
+    /// The solang parser failed
+    SolangParserFailed(Vec<Diagnostic>),
+    /// A Command Failed with error message
+    CommandFailed(String),
+    /// File IO Error
+    FileIoError(Box<dyn Error>),
+}
+
+impl ChiselDisptacher {
+    /// Associated public function to create a new Dispatcher instance
+    pub fn new(config: &SessionSourceConfig) -> Self {
+        ChiselDisptacher { errored: false, session: ChiselSession::new(config) }
+    }
+
+    /// Returns the prompt given the last input's error status
+    pub fn get_prompt(&self) -> String {
+        format!(
+            "{}{} ",
+            if let Some(id) = self.session.id {
+                format!("({}) ", format!("{}: {}", Paint::cyan("ID"), Paint::yellow(id)))
+            } else {
+                String::default()
+            },
+            if self.errored { Paint::red(PROMPT_ARROW) } else { Paint::green(PROMPT_ARROW) }
+        )
+    }
+
+    /// Dispatches a [ChiselCommand]
+    pub async fn dispatch_command(&mut self, cmd: ChiselCommand, args: &[&str]) -> DispatchResult {
+        match cmd {
+            ChiselCommand::Help => {
+                return DispatchResult::CommandSuccess(Some(format!(
+                    "{}\n{}",
+                    Paint::cyan(format!("{} Chisel help", CHISEL_CHAR)),
+                    ChiselCommand::iter()
+                        .map(|cmd| {
+                            let (cmd, desc) = CmdDescriptor::from(cmd);
+                            format!("!{} - {}", Paint::green(cmd), desc)
+                        })
+                        .collect::<Vec<String>>()
+                        .join("\n")
+                )))
+            }
+            ChiselCommand::Clear => {
+                if let Some(session_source) = self.session.session_source.as_mut() {
+                    // Drain all source sections
+                    session_source.drain_run();
+                    session_source.drain_global_code();
+                    session_source.drain_top_level_code();
+
+                    return DispatchResult::CommandSuccess(Some(String::from("Cleared session!")))
+                } else {
+                    return DispatchResult::CommandFailed(
+                        Paint::red("Session source not present!").to_string(),
+                    )
+                }
+            }
+            ChiselCommand::Flush => {
```

Comment:
Why `!flush`? I think `!save` is a bit more clear and intuitive 😅

Reply:
Updated :smile:

Reply:
Awesome, thanks! Related thought is that aliases would be nice, e.g. I keep trying to do `!h` instead of `!help`

Reply:
> Awesome, thanks! Related thought is that aliases would be nice, e.g. I keep trying to do `!h` instead of `!help`

Gotcha! Can add some shorthand aliases to the commands later on tonight.